### PR TITLE
Fix ambiguous column name that would prevent use of MSC2716 History Import when using Postgres as a database.

### LINKED ISSUE #12825: `TestImportHistoricalMessages` test fails in Complement when using workers
Interesting piece of log:

```
    msc2716_test.go:64: Deploy times: 22.33672842s blueprints, 7.183656158s containers
    msc2716_test.go:1138: Making POST request to http://localhost:49765/_matrix/client/r0/register (4682ca3c5471aa40266ac9893e1a927bbba68489636821bbb71c6b414bb8b107)
    msc2716_test.go:1138: Request body: {"type":"m.login.application_service","username":"maria"}
    client.go:544: POST hs1/_matrix/client/r0/register => 401 Unauthorized (1.838055ms)
    msc2716_test.go:1138: HTTP/1.1 401 Unauthorized
        Transfer-Encoding: chunked
        Access-Control-Allow-Headers: X-Requested-With, Content-Type, Authorization, Date
        Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
        Access-Control-Allow-Origin: *
        Cache-Control: no-cache, no-store, must-revalidate
        Connection: keep-alive
        Content-Type: application/json
        Date: Fri, 20 May 2022 16:26:35 GMT
        Server: nginx/1.18.0
        
        55
        {"errcode":"M_UNKNOWN_TOKEN","error":"Unrecognised access token","soft_logout":false}
        0
        
    msc2716_test.go:1157: msc2716.ensureVirtualUserRegistered failed to register: (M_UNKNOWN_TOKEN) Unrecognised access token
```
Obtained with
```
COMPLEMENT_DEBUG=1 WORKERS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestImportHistoricalMessages &> ../_misc/2022-05/compl-TestImportHistoricalMessages.log
```

Looks like it might just be a configuration problem (to add an appservice?)

### PR DESCRIPTION
**Pull request by reivilibre**
Fixes #12825.

@MadLittleMods I'd like to request your eyes to check this is what you intended.
It matches my reading of the surrounding commentary, but SQLite's query plan would have used `i` according to some experimentation!
Complement tests pass with *either* `c` or `i` so weren't much use in discerning the two :/.

### CODE REVIEW

File: synapse/storage/databases/main/event_federation.py [hunk 1]
```diff
@@ -1057,7 +1057,7 @@ def _get_connected_batch_event_backfill_results_txn(
             INNER JOIN batch_events AS c
             ON i.next_batch_id = c.batch_id
             /* Get the depth of the batch start event from the events table */
-            INNER JOIN events AS e USING (event_id)
+            INNER JOIN events AS e ON c.event_id = e.event_id
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:1060]: `c.event_id` (where `batch_events AS c`) makes sense. We want to return connected batch events and the info about the batch event (`depth`, `stream_ordering`)

Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:1060]: Can we configure SQLite to yell about ambiguous columns as well?

Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:1060]: > Complement tests pass with *either* `c` or `i` so weren't much use in discerning the two :/.

For reference, backfilling from the insertion or batch event will work. But we already have the insertion event so there isn't a need to backfill it again.

Comment (clokep) [reply synapse/storage/databases/main/event_federation.py:1060]: (Cross link please? 🙏 )

Comment (reivilibre) [reply synapse/storage/databases/main/event_federation.py:1060]: It's intentional behaviour and there are test cases ensuring it picks the first one: https://www.sqlite.org/forum/forumpost/941664fc450a4f56
Mostly as a legacy compatibility thing, but RIGHT and FULL joins will cause ambiguity checking (because old SQLites didn't support them and hence no applications can be relying on this).

File: changelog.d/12843.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix ambiguous column name that would prevent use of MSC2716 History Import when using Postgres as a database.
```
Comment (erikjohnston) [inline changelog.d/12843.bugfix:1]: Can you rephrase this as what the end user would see, rather than what is happening internally? I.e. something like "Fix bug where servers using SQLite3 would fail to backfill in MSC2716 rooms" or something.

The changelog is aimed at server admins rather than developers

**Review by erikjohnston — COMMENTED** [review]
Otherwise lgtm

File: changelog.d/12843.bugfix (hunk 2 above)
Comment (reivilibre) [reply changelog.d/12843.bugfix:1]: I think the problem is that I have never used this feature and don't really understand how this would look to a server admin :-).

> Fix bug where servers using a Postgres database would fail to import historical messages using MSC2716.

Would that be more suitable?

Comment (erikjohnston) [reply changelog.d/12843.bugfix:1]: Yeah, potentially? @MadLittleMods may have a better idea

Comment (MadLittleMods) [reply changelog.d/12843.bugfix:1]: It doesn't fail to import for me with my local Postgres Synapse setup.

Looking at this spot of code, I am guessing it will fail to backfill from another server when it sees an insertion event if you have `experimental_features.msc2716_enabled` enabled.

This seems more accurate:

> Fix bug where servers using a Postgres database would fail to backfill from an insertion event when [MSC2716](https://github.com/matrix-org/matrix-spec-proposals/pull/2716) is enabled (`experimental_features.msc2716_enabled`). 

---

```
$ psql -V
psql (PostgreSQL) 13.5

$ postgres -V
postgres (PostgreSQL) 13.5
```

Comment (reivilibre) [reply changelog.d/12843.bugfix]: Thanks, my message was purely going off the name of the test. What you've written makes sense.
